corehq/apps/translations/app_translations/upload_form.py

```diff
@@ -117,37 +116,54 @@ def update(self, rows):
-        #
+    def _update_or_create_translation_elements(self):
+        """
+        Create new elements if necessary and ensure existing elements are up to date with the current app config
+        """
+        template_element = self._get_template_translation_element()
+        for lang in self.langs:
+            translation_element = self.itext.find("./{f}translation[@lang='%s']" % lang)
+            if translation_element.exists():
+                self._update_default_attr_if_needed(translation_element, lang)
+            else:
+                self._create_translation_element(template_element, lang)
+
+    def _get_template_translation_element(self):
         # Currently operating under the assumption that every xForm has at least
-        # one translation element, that each translation element has a text node
-        # for each question and that each text node has a value node under it.
+        # one translation element, that each translation element has a text element
+        # for each question and that each text element has a value element under it.
         # Get a translation element to be used as a template for new elements, preferably of default lang
         default_lang = self.app.default_language
-        default_trans_el = self.itext.find("./{f}translation[@lang='%s']" % default_lang)
-        if default_trans_el.exists():
-            return default_trans_el
+        default_element = self.itext.find("./{f}translation[@lang='%s']" % default_lang)
+        if default_element.exists():
+            return default_element
+
+        # If no default element found, fallback to finding any translation element
         non_default_langs = copy.copy(self.app.langs)
         non_default_langs.remove(default_lang)
         for lang in non_default_langs:
-            trans_el = self.itext.find("./{f}translation[@lang='%s']" % lang)
-            if trans_el.exists():
-                return trans_el
-        raise Exception(_("Form has no translation node present to be used as a template."))
+            translation_element = self.itext.find("./{f}translation[@lang='%s']" % lang)
+            if translation_element.exists():
+                return translation_element
 
-    def _add_missing_translation_elements_to_itext(self, template_translation_el):
-        for lang in self.langs:
-            trans_el = self.itext.find("./{f}translation[@lang='%s']" % lang)
-            if not trans_el.exists():
-                new_trans_el = copy.deepcopy(template_translation_el.xml)
-                new_trans_el.set('lang', lang)
-                if lang != self.app.langs[0]:
-                    # If the language isn't the default language
-                    new_trans_el.attrib.pop('default', None)
-                else:
-                    new_trans_el.set('default', '')
-                self.itext.xml.append(new_trans_el)
+        raise Exception(_("Form has no translation element present to be used as a template."))
```

Comment:
I know you/we didn't add this, but is there a more specific exception that could be raised here?

Reply:
Sure. I checked the call stack and don't see anywhere that handles XFormExceptions specifically so I don't anticipate any behavior to change with this change and agree it is generally a good thing to do. 9dc05ba2c6c8ece942392597c063f663b2dd7d78